### PR TITLE
Improved Developer Documentation

### PR DESCRIPTION
**Pull request by opalmer**
This PR provides additional documentation for developers of pywincffi and clarifies a few practices of the project.

### CODE REVIEW

**Comment by opalmer** [comment]
@exvito, would you mind taking a look at this PR?

I've tried to document as much as I could around some of the things we discussed in #66 and wouldn't mind some feedback since you're the first person besides myself to work in pywincffi.  Also if you think there's something else that should be added I'd be more than happy to include it.

I plan on doing some more editing and may add some notes on the build process too but this PR is probably more or less complete otherwise.

File: README.rst [hunk 1]
```diff
@@ -111,6 +122,9 @@ typically used to reference the MSDN documentation but may also be used to
 reference examples, white papers or other reference which may be useful in
 describing the function.
 
+Adding new functions is covered in greater detail
+`here <https://https://pywincffi.readthedocs.org/en/latest/dev/functions.html>`_
```
Comment (exvito) [inline README.rst]: Duplicate `https://` in the link.

File: README.rst [hunk 2]
```diff
@@ -63,7 +74,7 @@ Documentation
 -------------
 
 The documentation for this this library is hosted on
-`Read The Docs <https://pywincffi.readthedocs.org/>`_.
+`ReadTheDocs <https://pywincffi.readthedocs.org/>`_.
```
Comment (exvito) [inline README.rst]: nit: Why no spaces? The homepage for it calls itself "Read the Docs", with spaces and with a non-capitalized "the".

File: docs/source/dev/codereview.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+told otherwise.  The master branch should be considered production ready and
+other branches are usually for testing and development.
+
+What Will Be Reviewed
+---------------------
+
+    * If a new function is being added, review the `function <functions.html>`_
+      documentation and make sure the new code matches these expectations.
+    * For style issues, the default rule of thumb is to follow PEP8 unless it's
+      something Windows specific then the `function <functions.html>`_
```
Comment (exvito) [inline docs/source/dev/codereview.rst]: Readability: The sentence could end in "unless it's something Windows specific" and be comma/semi-colon separated from the remaining "then ...".

File: docs/source/dev/functions.rst [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+   def AWindowsFunction(...):
+       """
+       First few sentences should tell someone what AWindowsFunction
+       does.  This can usually be pulled from the MSDN documentation but
+       is usually shorter and more concise.
+
+       .. seealso::
+
+          <url pointing to the msdn reference for AWindowsFunction>
+          <url pointing to a use case or other usefull information>
```
Comment (exvito) [inline docs/source/dev/functions.rst]: I'm not a native English speaker, but "useful" is written with a single "l", correct?

File: docs/source/dev/functions.rst [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+           always state key differences, if there are any, between what
+           the C api call normally expects and what the wrapper does.
+
+       <additional keyword or argument documentation>
+
+       :raises SomeException:
+           Information about under what condition(s) SomeException may be
+           raised.  SomeException should be something that's raised directly
+           by AWindowsFunction.
+
```
Comment (exvito) [inline docs/source/dev/functions.rst:263]: I saw a comment on existing code, which I used myself, describing `:rtype:` as well. Should it be here, if if optional?

File: docs/source/dev/functions.rst [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+      ffi.cast("DWORD", 0),
+      lpOverlapped
+   )
+
+
+Keywords
+~~~~~~~~
+In C, there's not really an equivalent to a keyword in Python.  However for
+many of the Windows API functions the msdn documentation may say something
+along the lines of *This parameter can be NULL.*  For pywincffi reasonable
```
Comment (exvito) [inline docs/source/dev/functions.rst]: Readability: I would add a "," between "For pywincffi" and "reasonable".

**Comment by exvito** [comment]
This is looking very good. I just pointed out one typo in a URL a few minor changes I would do but that come down as personal preference in writing style.

One aspect that is not explicitly covered within the context of adding functions, but that seems implicit from the codebase, is which module should a wrapper function be created in: currently there is one package -- kernel32 -- with a few modules, organized on a "per Windows object type" basis.

I assume the kernel32 name comes from the underlying kernel32.dll name. When adding a function in ntdll.dll or user32.dll should an appropriately named package be added? What about the related cdefs, should they go into the existing ones or should they be separated?

File: docs/source/dev/functions.rst (hunk 6 above)
Comment (opalmer) [inline docs/source/dev/functions.rst]: Done.

File: docs/source/dev/functions.rst (hunk 5 above)
Comment (opalmer) [inline docs/source/dev/functions.rst:263]: Yes, added.

File: docs/source/dev/functions.rst (hunk 4 above)
Comment (opalmer) [inline docs/source/dev/functions.rst]: Yes good catch thanks.

File: docs/source/dev/codereview.rst (hunk 3 above)
Comment (opalmer) [inline docs/source/dev/codereview.rst]: I think I got it, check 5339ee5 to see if that's what you meant.

File: README.rst (hunk 2 above)
Comment (opalmer) [inline README.rst]: Typo, not sure why it ended up this way.

File: README.rst (hunk 1 above)
Comment (opalmer) [inline README.rst]: Fixed.

**Comment by opalmer** [comment]
> One aspect that is not explicitly covered within the context of adding functions, but that seems implicit from the codebase, is which module should a wrapper function be created in: currently there is one package -- kernel32 -- with a few modules, organized on a "per Windows object type" basis.
> 
> I assume the kernel32 name comes from the underlying kernel32.dll name. When adding a function in ntdll.dll or user32.dll should an appropriately named package be added? 

Good question, I'll add this to the docs.

> What about the related cdefs, should they go into the existing ones or should they be separated?

Also a good question.  The short answer is, technically no they don't need to be and we don't need to care about it yet.  When and if we do need (or should) care about it changing it won't break backwards compatibility. I'll include some info in the docs about this too.

**Comment by opalmer** [comment]
Alright, commit eebe646 should have the remaining notes on C def/wrapper function locations.

**Comment by exvito** [comment]
This looks pretty good. I can't find any particular improvements to recommend.
Thanks.

**Comment by opalmer** [comment]
Thanks, merging.

**Pull request merged**